### DISCUSSION TITLE
Solving Poisson Equation with inf boundary condition

### DISCUSSION THREAD

**Discussion by sghelichkhani** [g-adopt/g-adopt#152] [Ideas]
Hi all! I am just sharing all the little I know about solving gravity using FEM here. This should be a good place for everyone to keep track of our progress and discussions. Also sort of getting @dsroberts started on this.

# Solving the Poisson Equation for Gravity Using Finite Elements

This document describes the theory and challenges associated with solving the Poisson equation for gravity using finite element methods (FEM), particularly within the Firedrake framework.

## The Poisson Equation for Gravity

The **Poisson equation** for gravitational potential is expressed as:

$$\nabla^2 \phi = 4\pi G \rho $$

and further computing gravity as

$$\mathbf{g} = - \nabla \phi$$

Where:

- $\phi$ is the gravitational potential,
- $G$ is the gravitational constant,
- $\rho$ is the mass density (which could be due to mass distribution of any form, including surface deflection)

## Analytical Solutions and Boundary Conditions

Analytical solutions to the Poisson equation are straightforward in specific cases:

- **Rectangular Domains:** Solutions often involve sine and cosine functions due to the harmonic nature of the equation in such geometries.
- **Spherical Domains:** Solutions typically utilise spherical harmonics, extending the concept of Fourier series to spherical geometries.

A primary challenge in applying FEM to the Poisson equation is handling boundary conditions. Theoretically, the boundary condition is set at infinity, where the gravitational potential \(\phi\) decays to zero. However, FEM requires a finite computational domain, necessitating boundary conditions at the truncated domain's edges. This limitation can hinder accurate approximation of the solution's behaviour at infinity.

## Coupling with Other Equations

The complexity increases when the Poisson equation is coupled with other systems, such as the Stokes equations(in case of mantle convection, and then for viscoelastic system is tiny bit different but very similar):

$$-\nabla P + \mu \nabla^2 \mathbf{u} + \rho \mathbf{g} = 0$$

$$\nabla \cdot \mathbf{u} = 0$$

Here:

- $\mathbf{g}$ represents gravitational acceleration, related to the gravitational potential by \(\mathbf{g} = -\nabla \phi\).

It's often assumed that the mass density $\rho$ influencing $\phi$ is independent of the Stokes equations' solution, implying the Poisson equation could be solved once without iterative updates (compute gravity once and then use it as an input parameter ). However, this is incorrect (There is a paper going through extensive math to prove that, which missed fundamental math definitions)!!! The density $\rho$ can include contributions from dynamic processes, such as non-spherical or horizontally varying deflections.

For instance, consider a dynamically deflected surface with topography ($h$), related to the radial stress $\sigma_{rr}$ and density contrast $\Delta \rho$ by:

$$h = \frac{\sigma_{rr}}{\Delta \rho g_0}$$

This topography introduces a mass distribution that generates a gravitational signal, altering the solution for \(\sigma_{rr}\). To account for this coupling, iterative methods like Picard iterations or non-linear Newton solvers should be employed to ensure consistency between the gravitational potential and the dynamic deflections.

## What should we do now?

Several methods have been proposed to address this in FEM. This is obviously what I have come across so far! 

### 1. Dirichlet-to-Neumann Operator

**Chaljub & Valette (2004)** use a Dirichlet-to-Neumann operator to derive a Robin-type boundary condition at a spherical boundary that is extended beyond the numerical domain we are interested in. This basically employs a coarse mesh extending far from the Earth and compute an analytical solution for the flux at that boundary. This approach allows setting a boundary condition combining both Dirichlet (value) and Neumann (flux) conditions, improving the approximation of the solution's behaviour at infinity.

### 2. Infinite Elements Approach

Used by **Gharti & Tromp (2017)** and **Gharti et al. (2018)**, they extend the computational domain virtually by mapping one side of the elements at the top boundary to infinity. Infinite elements adapt the test function space to accommodate the gravitational potential's decay with distance, effectively simulating an infinite domain. Our last discussion with David Ham when he was visiting us in 2022 implied that basically this was impossible at the time.

### 3. Multipole Expansion Method

**Van Driel et al. (2021)** present an approach similar to Chaljub & Valette but utilise a multipole expansion solution to obtain the longest wavelength components. This reduces computational effort by eliminating the need for iterative loops, requiring only one solution per time step. While this would be promising, I am not sure how "taped" all of these expansions would be for pyadjoint. 

### 4. Multiple approached by May Knepley 

**May & Knepley (2011)** introduce multiple approaches for computing gravity using FEM. TBH I have not really read this work. But I am sure this is an important work. Plus Matt's involvement with FD can help out a lot.

### 5. Absorbing Boundary Conditions in Firedrake

Recent developments by **Sagiyama & Ham (2023)** involve implementing absorbing boundary conditions (ABCs) for the Helmholtz equation. My understanding is that they also extend a mesh by using a "submesh", which they call absorbing layers. So basically  $Q$ finite element in conjunction with the N-point Gauss-Legendre quadrature reduced integration using Gauss-Legendre quadrature reduced integrations within Firedrake (You tell me what this means!). Since the Helmholtz equation is a Poisson equation with an additional term, their work may simplify implementing infinite boundary conditions in our case.

## Where to go next?
I honestly don't know which of these approaches work fine! The first try would be to use a script that I have from ages ago and see how far we should extend the mesh to able to have an acceptable solution. Koki's work seems to be most natural candidate because it's already in FD. van Driel's work is also sharing a lot of concepts. So that might be useful too. **The main two points: How to do all of that coupled to another system (Stokes/Richards Equations), and do not break taping by pyadjoint** 

---

## References

- Chaljub, E., & Valette, B. (2004). Spectral element modelling of three-dimensional wave propagation in a self-gravitating Earth with an arbitrarily stratified outer core. *Geophysical Journal International*, 158(1), 131–141. https://doi.org/10.1111/j.1365-246X.2004.02267.x

- Gharti, H. N., & Tromp, J. (2017). A spectral-infinite-element solution of Poisson’s equation: An application to self-gravity. *arXiv preprint arXiv:1706.00855v1*. https://doi.org/10.48550/arXiv.1706.00855

- Gharti, H. N., Tromp, J., & Zampini, S. (2018). Spectral-infinite-element simulations of gravity anomalies. *Geophysical Journal International*, 215(2), 1098–1117. https://doi.org/10.1093/gji/ggy469

- May, D. A., & Knepley, M. G. (2011). Optimal, scalable forward models for computing gravity anomalies. *Geophysical Journal International*, 187(1), 161–177. https://academic.oup.com/gji/article/187/1/161/560427

- Sagiyama, K., & Ham, D. A. (2023). Absorbing boundary conditions for the Helmholtz equation using Gauss-Legendre quadrature reduced integrations. *arXiv preprint arXiv:2308.12255*. https://doi.org/10.48550/arXiv.2308.12255

- Van Driel, M., Kemper, J., & Boehm, C. (2021). On the modelling of self-gravitation for full 3-D global seismic wave propagation. *Geophysical Journal International*, 227(1), 632–643. https://doi.org/10.1093/gji/ggab237

**Comment by w3168** [comment]
Thanks for sharing this @sghelichkhani. This is the example code snippet I mentioned in the pub yesterday re the coupling problem. https://www.firedrakeproject.org/firedrake.html#firedrake.mesh.Submesh 

When I saw it a few weeks ago I expected it to work...! But it looks like this functionality for defining a mixed function space on two meshes (or submeshes...) isn't fully integrated yet. Using a newish Firedrake build from last week

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[1], line 12
     10 V0 = FunctionSpace(mesh, "CG", 1)
     11 V1 = FunctionSpace(subm, "CG", 1)
---> 12 V = V0 * V1
     13 u = TrialFunction(V)
     14 v = TestFunction(V)

File ~/firedrake/src/firedrake/firedrake/functionspaceimpl.py:327, in WithGeometryBase.__mul__(self, other)
    324 r"""Create a :class:`.MixedFunctionSpace` composed of this
    325 :class:`.FunctionSpace` and other"""
    326 from firedrake.functionspace import MixedFunctionSpace
--> 327 return MixedFunctionSpace((self, other))

File petsc4py/PETSc/Log.pyx:188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func()

File petsc4py/PETSc/Log.pyx:189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func()

File ~/firedrake/src/firedrake/firedrake/functionspace.py:280, in MixedFunctionSpace(spaces, name, mesh)
    278 for i in range(1, len(meshes)):
    279     if meshes[i] is not meshes[0]:
--> 280         raise ValueError("All function spaces must be defined on the same mesh!")
    282 try:
    283     cls, = set(type(s) for s in spaces)

ValueError: All function spaces must be defined on the same mesh!
```

Anyway I did a tiny bit more playing around this afternoon and the code snippet now runs with this branch https://github.com/firedrakeproject/firedrake/pull/3478 (and the associated ufl and tsfc branches).  I added a few print statements and it looks like it is integrating the parts of the domain correctly... 
```
print(A)

b = assemble(Constant(1)*dx0)
print(b)

c = assemble(Constant(1)*dx1)
print(c)
```
gives
```
assembled Matrix(a={ v_0[0] * v_1[1] } * dx(<Mesh #28>[999], {})
  +  { v_1[0] * v_0[1] } * dx(<Mesh #29>[everywhere], {}), bcs=())
1.9999999999999987
0.9999999999999993
```



I guess at the moment we could probably solve gravity and stokes explicitly on Firedrake's master branch (maybe with a handcoded picard loop) just by interpolating the gravity field for the large mesh to the submesh for the mantle domain and vice versa, which I think got introduced in this PR https://github.com/firedrakeproject/firedrake/pull/3727 

And then if that works we can probably build on the internal variable coupling to try and solve stokes+gravity as a coupled system

**Comment by w3168** [comment]
@MorrowL @sghelichkhani https://arxiv.org/pdf/2009.08493
